### ISSUE TITLE
Strange interaction of glob rules prevents rules from firing

### ISSUE THREAD

**Issue by parsonsmatt** [ndmitchell/hlint#1647]
We have two rules:

```haskell
- modules:
    - name:
          - "**.*Spec"
      within: 
          - Main
          - Spec
    - name:
          - "Handler.**"
      within: 
          - "Handler.**"
          - "**.*Spec"
```

The intention of these rules are:

1. You should not import `Spec` modules except in defining tests. Shared helpers should be extracted to utility modules.
2. You should not import `Handler` modules except in `Handler` code and in tests for `Handler` code.

However, the rules appear to be merged. We recently discovered that rules matching the first were not being triggered except in pre-commit, which was only loading that rule. However, these rules were not being enforced in CI or in our `make hlint` or `make hlint-all` commands, which loaded all of the rules. 

I am working up a minimal reproduction of this behavior.

**Comment by parsonsmatt** [comment]
OK, I have a reproduction over here: https://github.com/parsonsmatt/hlint-repro-1647

I need to figure out if there's more going on, but at least one problematic behavior: when an import matches two rules, then the import will be allowed if *either* rule has a `within` exclusion match. The above reproduction has two rules:

```yaml
- modules:
    - name: "Handler.**"
      within: 
        - "Handler.**"
        - "**.*Spec"
    - name: "**.*Spec"
      within:
        - Spec 
```

This file and import should report an error due to a violation of the rule `**.*Spec`.

```haskell
module Handler.FooSpec where

import Handler.BarSpec
```

However, becaues the import *also* matches the rule for `Handler.**`, which is *allowed* in `**.*Spec` and also allowed in `Handler.**`, this prevents the rule for "import **.*Spec" from firing.
